controllers/user/controllers/helper/csr.go

```diff
@@ -0,0 +1,224 @@
+	"k8s.io/apimachinery/pkg/watch"
+	csrv1client "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+)
+
+// newPrivateKey creates an RSA private key
+func newPrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
+	if keyType == x509.ECDSA {
+		return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	}
+
+	return rsa.GenerateKey(rand.Reader, 2048)
+}
+
+func NewSignedToCsrKey(cfg Config) (csr, keyPEM []byte, err error) {
+	key, err := newPrivateKey(x509.RSA)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new signed private failed %s", err)
+	}
+	pk := x509.MarshalPKCS1PrivateKey(key.(*rsa.PrivateKey))
+	keyPEM = pem.EncodeToMemory(&pem.Block{
+		Type: "RSA PRIVATE KEY", Bytes: pk,
+	})
+	_, csr, err = generateCSRFile(cfg, key)
+
+	csr = pem.EncodeToMemory(&pem.Block{
+		Type: "CERTIFICATE REQUEST", Bytes: csr,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("new signed csr failed %s", err)
+	}
+
+	return
+}
+
+type TLS struct {
+	Name   string
+	CAKey  []byte
+	TLSKey []byte
+	TLSCrt []byte
+	TLSCsr []byte
+}
+
+func GenerateKubeConfigFromCsr(cfg Config) ([]byte, error) {
+	if cfg.ClusterName == "" {
+		cfg.ClusterName = "kubernetes"
+	}
+	csr, key, err := NewSignedToCsrKey(cfg)
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewKubernetesClient("", "")
+	if err != nil {
+		return nil, err
+	}
+	// make sure cadata is loaded into config under incluster mode
+	if err = rest.LoadTLSFiles(client.Config()); err != nil {
+		return nil, err
+	}
+	ca := client.Config().CAData
+	tls := &TLS{}
+	tls.TLSCsr = csr
+	tls.TLSKey = key
+	tls.CAKey = ca
+	tls.Name = cfg.User
+	if err = updateCsr(tls, client.Kubernetes().CertificatesV1().CertificateSigningRequests()); err != nil {
+		return nil, err
+	}
+	ctx := fmt.Sprintf("%s@%s", cfg.User, cfg.ClusterName)
+	config := &api.Config{
+		Clusters: map[string]*api.Cluster{
+			cfg.ClusterName: {
+				Server:                   client.Config().Host,
+				CertificateAuthorityData: ca,
+			},
+		},
+		Contexts: map[string]*api.Context{
+			ctx: {
+				Cluster:  cfg.ClusterName,
+				AuthInfo: cfg.User,
+			},
+		},
+		AuthInfos: map[string]*api.AuthInfo{
+			cfg.User: {
+				ClientCertificateData: tls.TLSCrt,
+				ClientKeyData:         tls.TLSKey,
+			},
+		},
+		CurrentContext: ctx,
+	}
+	kubeData, err := clientcmd.Write(*config)
+	if err != nil {
+		return nil, fmt.Errorf("get kubeconfig failed %s", err)
+	}
+
+	return kubeData, nil
+}
+
+func updateCsr(tls *TLS, client csrv1client.CertificateSigningRequestInterface) error {
+	dPolicy := v1.DeletePropagationBackground
+	csrName := fmt.Sprintf("sealos-generater-%s", tls.Name)
+	label := map[string]string{
+		"csr-name": csrName,
+	}
+	_ = client.Delete(context.TODO(), csrName, v1.DeleteOptions{PropagationPolicy: &dPolicy})
+	csrResource := &csrv1.CertificateSigningRequest{}
+	seconds := int32(86400)
```

Comment:
The expired time currently is 24h, witch seems to be to long, may be we can expose it to configureable for differ perpose.(test/prod have diferrent expire time)